File: M1files/lexer.cpp
#include <iostream>
#include <string>
#include <regex>

#include "HeaderFiles/tokens.h"
#include "HeaderFiles/lexer.h"

Lexer::Lexer(std::string &source) {
    _source = source;
    _character = '\0';
    _position = 0;
    _read_position = 0;
}

void Lexer::_read_character() {
    if (_read_position >= _source.size()) {
        _character = '\0';
    } else {
        _character = _source[_read_position];
    }
    _position = _read_position;
    _read_position++;
}

char Lexer::_peek_character() {
    if (_read_position >= _source.size()) {
        return '\0';
    }
    return _source[_read_position];
}

Token Lexer::_make_two_character_token(TokenType token_type) {
    char prefix = _character;
    _read_character();
    char suffix = _character;
    return Token(token_type, std::string(1, prefix) + suffix);
}

void Lexer::_skip_whitespace() {
    while (std::regex_match(std::string(1, _character), std::regex("^\\s$"))) {
        _read_character();
    }
}

bool Lexer::_is_letter(char character) {
    return std::regex_match(std::string(1, character), std::regex("^[a-zA-ZáéíóúÁÉÍÓÚñÑ_]$"));
}

bool Lexer::_is_number(char character) {
    return std::regex_match(std::string(1, character), std::regex("^\\d$"));
}

std::string Lexer::_read_number() {
    int initial_position = _position;
    while (_is_number(_character)) {
        _read_character();
    }
    return _source.substr(initial_position, _position - initial_position);
}

std::string Lexer::_read_identifier() {
    int initial_position = _position;
    while (_is_letter(_character)) {
        _read_character();
    }
    return _source.substr(initial_position, _position - initial_position);
}

Token Lexer::next_token() {
    _skip_whitespace();
    Token token(TokenType::ILLEGAL, std::string(1, _character)); // Default to ILLEGAL token

    if (_character == '=') { // Assign = or Equals ==
        if (_peek_character() == '=') {
            token = _make_two_character_token(TokenType::EQ);
        } else {
            token = Token(TokenType::ASSING, std::string(1, _character));
        }
    } else if (_character == '!') { // Negation ! or Dif !=
        if (_peek_character() == '=') {
            token = _make_two_character_token(TokenType::DIF);
        } else {
            token = Token(TokenType::NEGATION, std::string(1, _character));
        }
    } else if (_character == '+') { // Plus +
        token = Token(TokenType::PLUS, std::string(1, _character));
    } else if(_character == '{') { // Lbrace {
        token = Token(TokenType::LBRACE, std::string(1, _character));
    } else if(_character == '}') { // Rbrace }
        token = Token(TokenType::RBRACE, std::string(1, _character));
    } else if(_character == '(') { // Lparen (
        token = Token(TokenType::LPAREN, std::string(1, _character));
    } else if(_character == ')') { // Rparen )
        token = Token(TokenType::RPAREN, std::string(1, _character));
    } else if(_character == '-') { // Minus -
        token = Token(TokenType::MINUS, std::string(1, _character));
    } else if(_character == ',') { // Comma ,
        token = Token(TokenType::COMMA, std::string(1, _character));
    } else if(_character == ';') { // Semicolon ;
        token = Token(TokenType::SEMICOLON, std::string(1, _character));
    } else if (_is_letter(_character)) { // Letter
        std::string literal = _read_identifier();
        TokenType token_type = lookup_token_type(literal);
        return Token(token_type, literal);
    } else if (_is_number(_character)) { // Number
        std::string literal = _read_number();
        TokenType token_type = lookup_token_type(literal);
        return Token(TokenType::INTEGER, literal);
    } else if(_character == '>') { // Greater than > or equals >=
        if(_peek_character() == '=')
            token = _make_two_character_token(TokenType::GTE);
        else
            token = Token{TokenType::GT, std::string(1, _character)};
    } else if(_character == '<') // Less than < or equals <=
    {
        if(_peek_character() == '=')
            token = _make_two_character_token(TokenType::LTE);
        else
            token = Token{TokenType::LT, std::string(1, _character)};
    } else if(_character == '\0') { // End of file
        token = Token(TokenType::EOF_TOKEN, std::string(1, _character));
    }
    _read_character();
    return token;
}

File: M1files/Lexer.java
package M1files;

public class Lexer {
    private String source;
    private char character;
    private int position;
    private int readPosition;

    public Lexer(String source) {
        this.source = source;
        this.position = 0;
        this.readPosition = 0;
        readCharacter();
    }

    public Token nextToken(){
        skipWhitespace();
        Token token = new Token(TokenType.ILLEGAL, String.valueOf(character));
        if(character == '=') {
            if(peekCharacter() == '=') { // Assing or equals '=' or '=='
                token =  makeTwoCharacterToken(TokenType.EQ); // '=='
            } else {
                token = new Token(TokenType.ASSIGN, String.valueOf(character)); // '='
            }
        } else if(character == ',') { // Comma ,
            token = new Token(TokenType.COMMA, String.valueOf(character));
        } else if(character == '!') { // Different or equals '!='
            if(peekCharacter() == '=') { 
                token =  makeTwoCharacterToken(TokenType.DIF); // '!='
            } else {
                token =  new Token(TokenType.NEGATION, String.valueOf(character)); // '!'
            }
        } else if(character == '/') { // Division
            token = new Token(TokenType.DIVISION, String.valueOf(character));
        } else if(character == '\u0000') { // Blank
            token = new Token(TokenType.EOF, String.valueOf(character));
        } else if(Character.isLetter(character)) { // Letter
            String literal = readIdentifier();
            TokenType tokenType = Translator.lookupTokenType(literal);
            token = new Token(tokenType, literal);
        } else if(character == '>') { // Greater than 
            if(peekCharacter() == '=') {
                token = makeTwoCharacterToken(TokenType.GTE); // '>='
            } else {
                token = new Token(TokenType.GT, String.valueOf(character)); // '>'
            }
        } else if(Character.isDigit(character)){ // Number
            String literal = readNumber();
            //TokenType tokenType = Translator.lookupTokenType(literal);
            token = new Token(TokenType.INTEGER, literal);
        } else if(character == '{') { // Rbrace '{'
            token = new Token(TokenType.LBRACE, String.valueOf(character));
        } else if(character == '(') { // Lparen '('
            token = new Token(TokenType.LPAREN, String.valueOf(character));
        } else if(character == '<') { // Less than
            if(peekCharacter() == '=') {
                token = makeTwoCharacterToken(TokenType.LTE); // '<='
            } else {
                token = new Token(TokenType.LT, String.valueOf(character)); // '<'
            }
        } else if(character == '-') { // Minus '-'
            token = new Token(TokenType.MINUS, String.valueOf(character));
        } else if(character == '*') { // Multiplication '*'
            token = new Token(TokenType.MULTIPLICATION, String.valueOf(character));
        } else if(character == '+') { // Plus '+'
            token = new Token(TokenType.PLUS, String.valueOf(character));
        } else if(character == '}'){ // Rbrace '}'
            token = new Token(TokenType.RBRACE, String.valueOf(character));
        } else if(character == ')'){ // Rparen ')'
            token = new Token(TokenType.RPAREN, String.valueOf(character));
        } else if(character == ';') { // Semicolon ';'
            token = new Token(TokenType.SEMICOLON, String.valueOf(character));
        }
        readCharacter();
        return token;
    }

    private String readNumber(){
        int initialPosition = position;
        while(Character.isDigit(character)) {
            readCharacter();
        }
        return source.substring(initialPosition, position);
    }
    private String readIdentifier(){
        int initialPosition = position;
        while(position < source.length() && Character.isLetter(character)) {
            readCharacter();
        }
        return source.substring(initialPosition, position);
    }
    private void readCharacter(){
        if(readPosition >= source.length()) {
            character = '\u0000';
        } else {
            character = source.charAt(readPosition);
        }
        position = readPosition;
        readPosition++;
    }
    private char peekCharacter(){
        if(readPosition >= source.length()) {
            return '\u0000';
        } 
        return source.charAt(readPosition);
    }
    private Token makeTwoCharacterToken(TokenType tokenType){
        char prefix = character;
        readCharacter();
        char suffix = character;
        return new Token(tokenType, String.valueOf(prefix) + suffix);
    }
    private void skipWhitespace(){
        while(character == ' ' || character == '\t' || character == '\n' || character == '\r') {
            readCharacter();
        }
    }
}